File: mqtt-blinky/user/user_data.nim
import esp8266/crc
import esp8266/types


const
  magic = 0x0A0CDCBA'u32
  app = "mqbl"


type
  WiFiSsid* = array[32, uint8]
  WiFiPassword* = array[64, uint8]
  WiFiCreds* {.packed.} = object
    ssid*: WiFiSsid
    password*: WiFiPassword
  Mqtt* {.packed.} = object
    ip*: array[4, uint8]
    port*: uint16
    username* {.align(4).}: array[64, uint8]
    password* {.align(4).}: array[64, uint8]
  Settings* {.packed.} = object
    magic*: uint32
    app: array[4, uint8]
    wifi*: WifiCreds
    mqtt*: Mqtt
  Data* {.packed.} = object
    settings*: Settings
    crc32*: uint32


proc verify*(data: var Data): (bool, string) =
  if data.settings.magic != magic:
    return (false, "wrong magic")
  if data.settings.app != app.toArray(4, uint8):
    return (false, "wrong app")
  let crc = crc32(data.settings)
  if data.crc32 != crc:
    return (false, "bad crc")
  return (true, "")


when isMainModule:
  import endians
  import net
  import parseopt
  import streams
  import strutils

  import esp8266/types

  var
    data: Data
    ssid: string
    wifiPassword: string
    mqttUsername: string
    mqttPassword: string
    host: string
    dest = "user_data.bin"
    optParser = initOptParser()

  while true:
    optParser.next()
    case optParser.kind
    of cmdEnd: break
    of cmdShortOption, cmdLongOption:
      case optParser.key:
      of "ssid", "s":
        ssid = optParser.val
      of "password", "passwd", "pass", "p":
        wifiPassword = optParser.val
      of "host":
        host = optParser.val
      of "mqtt-username":
        mqttUsername = optParser.val
      of "mqtt-password":
        mqttPassword = optParser.val
      else:
        echo("Unknown option: ", optParser.key)
        quit(QuitFailure)
      if optParser.val == "":
        echo("Option: ", optParser.key, " takes one argument")
        quit(QuitFailure)
    of cmdArgument:
      dest = optParser.key

  if ssid == "":
    stdout.write("SSID: ")
    ssid = readLine(stdin)
  if wifiPassword == "":
    stdout.write("WiFi Password: ")
    wifiPassword = readLine(stdin)
  if host == "":
    let default_host = $getPrimaryIPAddr() & ":1883"
    stdout.write("Using the format: IPv4:PORT\n")
    stdout.write("Leave blank for: " & default_host & "\n")
    stdout.write("Host: ")
    host = readLine(stdin)
    if host == "":
      host = default_host
  if mqttUsername == "":
    stdout.write("MQTT Username (blank for none): ")
    mqttUsername = readLine(stdin)
  if mqttPassword == "":
    stdout.write("MQTT Password (blank for none): ")
    mqttPassword = readLine(stdin)

  var parts = host.split(':', 1)
  let ip = parseIpAddress(parts[0])
  let port = parts[1].parseUInt()

  data.settings.magic = magic
  data.settings.app.setString(app)
  data.settings.wifi.ssid.setString(ssid)
  data.settings.wifi.password.setString(wifiPassword)
  data.settings.mqtt.ip = ip.address_v4
  data.settings.mqtt.port = uint16(port)
  data.settings.mqtt.username.setString(mqttUsername)
  data.settings.mqtt.password.setString(mqttPassword)
  data.crc32 = crc32(data.settings)

  littleEndian32(addr data.settings.magic, addr data.settings.magic)
  littleEndian32(addr data.settings.mqtt.port, addr data.settings.mqtt.port)
  littleEndian32(addr data.crc32, addr data.crc32)

  let padding = '\xff'.repeat(4096 - sizeof(data))

  let fp = open(dest, fmWrite)
  discard fp.writeBuffer(addr data, sizeof(data))
  fp.write(padding)
  fp.close()

File: mqtt-blinky/user_data.py
# A cross platform Python script to generate a user_data.bin file
# for the FOTA example.
import binascii
import struct
import sys
import socket

try:
    input = raw_input
except NameError:
    pass


def get_primary_ip():
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        s.connect(('10.255.255.255', 1))  # doesn't have to be reachable
        ip = s.getsockname()[0]
    except Exception:
        ip = '127.0.0.1'
    finally:
        s.close()
    return ip


default_host = get_primary_ip() + ':1883'

magic = 0x0a0cdcba
app = 'mqbl'
ssid = input('SSID: ').encode('ascii')
password = input('Password: ').encode('ascii')
print('Using the format: IPv4:PORT')
print('Leave blank for: ' + default_host)
host = input('Host: ')
if not host:
    host = default_host
mqtt_username = input('MQTT Username (leave blank for none): ').encode('ascii')
mqtt_password = input('MQTT Password (leave blank for none): ').encode('ascii')


ip, port = host.split(':', 1)
ip = socket.inet_aton(ip)
port = int(port)


filename = sys.argv[1] if len(sys.argv) > 1 else '0x70000.bin'
with open(filename, 'wb') as fp:
    settings = struct.pack('<I4s32s64s4sHxx64s64s',
                           magic, app, ssid, password, ip,
                           port, mqtt_username, mqtt_password)
    crc = binascii.crc32(settings) & 0xffffffff
    data = settings + struct.pack('<I', crc)
    data += b'\xff' * (4096 - len(data))
    fp.write(data)
